### PR TITLE
Added approximations

### PR DESCRIPTION
**Pull request by aanurraj**
## Description
Added approximation functions

## Affected Dependencies
none

## How has this been tested?
added tests

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests
- [x] pow
- [x] exp
- [x] log
- [x] reciprocal
- [x] sigmod
- [x] tanh

### CODE REVIEW

**Comment by gmuraru** [comment]
I think it looks ok for a first implementation.

Could you also add tests for them? :D

File: src/sympc/tensor/mpc_tensor.py [hunk 1]
```diff
@@ -438,6 +438,37 @@ def div(self, y: Union["MPCTensor", torch.Tensor, float, int]) -> "MPCTensor":
         result = spdz.public_divide(self, y)
         return result
 
+    def pow(self, power: int) -> "MPCTensor":
+        """Compute integer power of a number by recursion using mul This uses
```
Comment (gmuraru) [inline src/sympc/tensor/mpc_tensor.py:442]: TYPO: Not finished

File: src/sympc/approximations/sigmoid.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+def sigmoid(tensor, method):
+    """
+    Approximates the sigmoid function using a given method
+    Args:
+        tensor: the fixed precision tensor
+        method (str): (default = "chebyshev")
+            Possible values: "exp", "maclaurin", "chebyshev"
+    """
+    if method == "exp":
+        _sign = sign(tensor)
```
Comment (gmuraru) [inline src/sympc/approximations/sigmoid.py:18]: Q: Could you add a dictionary with different functions for this.

File: src/sympc/approximations/reci.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                :math:`x^{-1} = exp(-log(x))`
+        nr_iters:
+            Number of iterations for `Newton-Raphson`
+    Returns:
+        Reciprocal of `self`
+    """
+    method = method.lower()
+
+    if method == "nr":
+        new_self = modulus(self)
```
Comment (gmuraru) [inline src/sympc/approximations/reci.py:27]: Also this

File: tests/sympc/approximations/exponential_test.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    SessionManager.setup_mpc(session_one)
+
+    x_secret = torch.Tensor([0.0, 1, 2, 3, 4])
+
+    # default precision
+    x = MPCTensor(secret=x_secret, session=session_one)
+    x_exp = exp(x)
+    x_secret_exp = torch.exp(x_secret)
+
+    assert torch.allclose(x_secret_exp, x_exp.reconstruct(), rtol=0.4)
```
Comment (gmuraru) [inline tests/sympc/approximations/exponential_test.py:23]: Q: Can we remove the ```rtol```? (like it throws errors with it?)

File: tests/sympc/approximations/log_test.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    SessionManager.setup_mpc(session_one)
+
+    x_secret = torch.Tensor([0.1, 0.5, 2, 5, 10, 20, 50, 100, 250])
+
+    # default precision
+    x = MPCTensor(secret=x_secret, session=session_one)
+    x_log = log(x)
+    x_secret_log = torch.log(x_secret)
+
+    assert torch.allclose(x_secret_log, x_log.reconstruct(), rtol=0.4)
```
Comment (gmuraru) [inline tests/sympc/approximations/log_test.py:23]: rtol should be somewhere as ```1e-3``` or ```1e-4```. 0.4 is a big value

Comment (gmuraru) [reply tests/sympc/approximations/log_test.py:23]: For all the cases

File: tests/sympc/approximations/reci_test.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    clients = get_clients(2)
+    session_one = Session(parties=clients)
+    SessionManager.setup_mpc(session_one)
+
+    x_secret = torch.Tensor([-2.0, 6.0, 2.0, 3.0, -5.0, -0.5])
+
+    x = MPCTensor(secret=x_secret, session=session_one)
+    x_secret_reciprocal = torch.reciprocal(x_secret)
+
+    x_reciprocal = reciprocal(x, method="nr")
```
Comment (gmuraru) [inline tests/sympc/approximations/reci_test.py:20]: MAYBE: Use the ```method``` name with ```pytest.parametrize```

File: tests/sympc/approximations/exponential_test.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+from sympc.session import SessionManager
+from sympc.tensor.mpc_tensor import MPCTensor
+
+
+def test_exp(get_clients) -> None:
+    clients = get_clients(2)
+    session_one = Session(parties=clients)
+    SessionManager.setup_mpc(session_one)
+
+    x_secret = torch.Tensor([0.0, 1, 2, 3, 4])
```
Comment (gmuraru) [inline tests/sympc/approximations/exponential_test.py:16]: Q: Could you also add some negative values in this tensor?

File: src/sympc/approximations/utils.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+def sign(data):
```
Comment (gmuraru) [inline src/sympc/approximations/utils.py:1]: Q: I am not sure what is the difference between ```signum``` and ```sign```

File: tests/sympc/approximations/reci_test.py (hunk 6 above)
Comment (aanurraj) [reply tests/sympc/approximations/reci_test.py:20]: okay

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 9]
```diff
@@ -363,3 +363,20 @@ def test_share_get_method_session(get_clients) -> None:
...
 
     assert all(res.get() == expected_res)
+
+
+def test_pow(get_clients) -> None:
+    clients = get_clients(2)
+    session = Session(parties=clients)
+    SessionManager.setup_mpc(session)
+
+    x_secret = torch.Tensor([5.0])
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:373]: MAYBE: Also add a negative value and parametrize this test to raise the tensor to an odd and even power

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 10]
```diff
@@ -363,3 +363,20 @@ def test_share_get_method_session(get_clients) -> None:
     res = mpc_tensor * mpc_tensor
 
     assert all(res.get() == expected_res)
+
+
+def test_pow(get_clients) -> None:
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:368]: MAYBE: Also add a test where you check the ```Exception``` in case the power is negative

File: .gitignore [hunk 11]
```diff
@@ -52,3 +52,4 @@ MANIFEST
 
 # VSCode
 .vscode/
+/venv/
```
Comment (gmuraru) [inline .gitignore:55]: Q: is this needed?

**Review by gmuraru — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 11 above)
Comment (aanurraj) [reply .gitignore:55]: ahh might be my IDE.. removing now

File: tests/sympc/approximations/exponential_test.py (hunk 4 above)
Comment (aanurraj) [reply tests/sympc/approximations/exponential_test.py]: opened an issue regarding this

File: tests/sympc/approximations/log_test.py (hunk 5 above)
Comment (aanurraj) [reply tests/sympc/approximations/log_test.py]: updated

File: tests/sympc/approximations/exponential_test.py (hunk 7 above)
Comment (aanurraj) [reply tests/sympc/approximations/exponential_test.py]: removed exp test and opened an issue

File: tests/sympc/tensor/mpc_tensor_test.py (hunk 9 above)
Comment (aanurraj) [reply tests/sympc/tensor/mpc_tensor_test.py]: done

File: .gitignore (hunk 11 above)
Comment (aanurraj) [reply .gitignore]: done

File: src/sympc/approximations/utils.py (hunk 8 above)
Comment (aanurraj) [reply src/sympc/approximations/utils.py:4]: ahh it was redundant removed it we now have only sign

File: src/sympc/tensor/mpc_tensor.py (hunk 1 above)
Comment (aanurraj) [reply src/sympc/tensor/mpc_tensor.py]: done

**Review by gmuraru — APPROVED** [review]
LGTM!
